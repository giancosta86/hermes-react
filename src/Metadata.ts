import { RMap } from "@rimbu/collection-types";

export type Char = string;

export type CharMetadata = Readonly<{
  annotation?: string;
  className?: string;
}>;

export type CharMetadataMap = RMap<Char, CharMetadata>;

export type MetadataByChar = Readonly<{ [char: Char]: CharMetadata }>;
