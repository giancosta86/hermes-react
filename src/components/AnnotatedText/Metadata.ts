import { RMap } from "@rimbu/collection-types";

export type Char = string;

export type CharMetadata = Readonly<{
  annotation: string;
  classes: Iterable<string>;
}>;

export type MetadataByChar = RMap<Char, CharMetadata>;
