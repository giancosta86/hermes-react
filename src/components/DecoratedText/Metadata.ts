import { RMap } from "@rimbu/collection-types";

export type Char = string;

export type CharMetadata = Readonly<{
  annotation?: string;
  className?: string;
}>;

export type MetadataByChar = RMap<Char, CharMetadata>;
