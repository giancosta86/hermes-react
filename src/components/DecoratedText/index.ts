export * from "./DecoratedText";

export * from "./Metadata";
