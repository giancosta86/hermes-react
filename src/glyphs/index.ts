export * as Glyph from "./Glyph";

export * as Pinyin from "./Pinyin";

export * as PinyinTone from "./PinyinTone";

export * as Russian from "./russian";
