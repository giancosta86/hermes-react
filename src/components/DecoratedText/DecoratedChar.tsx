import React from "react";
import { CharMetadata } from "@/Metadata";

export interface DecoratedCharProps {
  char: string;
  metadata: CharMetadata;
}

export const DecoratedChar = ({ char, metadata }: DecoratedCharProps) =>
  metadata.annotation ? (
    <ruby className={metadata.className}>
      {char}
      <rp>(</rp>
      <rt>{metadata.annotation}</rt>
      <rp>)</rp>
    </ruby>
  ) : (
    <span className={metadata.className}>{char}</span>
  );
