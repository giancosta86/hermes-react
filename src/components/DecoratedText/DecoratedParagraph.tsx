import React, { useMemo } from "react";
import { MetadataByChar } from "./Metadata";
import { DecoratedChar } from "./DecoratedChar";

export interface DecoratedParagraphProps {
  text: string;
  metadataByChar: MetadataByChar;
}

export const DecoratedParagraph = ({
  text,
  metadataByChar
}: DecoratedParagraphProps) => {
  const chars = useMemo(() => text.split(""), [text]);

  return (
    <>
      {chars.map((char, index) => {
        if (char == "\n") {
          return <br key={index} />;
        }

        const metadata = metadataByChar.get(char);

        return metadata && (metadata.annotation || metadata.className) ? (
          <DecoratedChar key={index} char={char} metadata={metadata} />
        ) : (
          <React.Fragment key={index}>{char}</React.Fragment>
        );
      })}
    </>
  );
};
