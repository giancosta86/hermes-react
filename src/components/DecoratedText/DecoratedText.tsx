import React, { useMemo } from "react";
import { MetadataByChar } from "./Metadata";
import { DecoratedParagraph } from "./DecoratedParagraph";

export interface DecoratedTextProps {
  text: string;
  metadataByChar: MetadataByChar;
}

export const DecoratedText = ({ text, metadataByChar }: DecoratedTextProps) => {
  const paragraphs = useMemo(
    () =>
      text
        .trim()
        .replaceAll("\r\n", "\n")
        .split(/\n\n+/)
        .filter(line => line.length),
    [text]
  );

  return (
    <>
      {paragraphs.map((paragraph, index) => (
        <p key={index}>
          <DecoratedParagraph
            text={paragraph}
            metadataByChar={metadataByChar}
          />
        </p>
      ))}
    </>
  );
};
