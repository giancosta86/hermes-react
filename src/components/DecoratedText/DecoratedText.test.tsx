import React from "react";
import { create } from "react-test-renderer";
import { DecoratedText } from "./DecoratedText";
import { HashMap } from "@rimbu/hashed";
import { Char, CharMetadata } from "./Metadata";
import { Pinyin } from "@/glyphs";

describe("<DecoratedText>", () => {
  describe("when receiving text with no newlines", () => {
    it("should just render it in a single paragraph", () => {
      const renderer = create(
        <DecoratedText
          text={"我看书"}
          metadataByChar={HashMap.of<Char, CharMetadata>([
            "书",
            { annotation: "shū", className: Pinyin.flat }
          ])}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving text with leading and trailing space and no metadata", () => {
    it("should trim the text and render just plain characters in a single paragraph", () => {
      const renderer = create(
        <DecoratedText
          text={"  \t \nDodo  \r\n  \t"}
          metadataByChar={HashMap.empty()}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving text with multiple intermediate carriageReturn + lineFeed", () => {
    it("should ignore carriage returns", () => {
      const renderer = create(
        <DecoratedText
          text={"Alpha\r\nBeta\r\nGamma"}
          metadataByChar={HashMap.empty()}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving text with long sequences of intermediate newlines", () => {
    it("should ignore additional newlines", () => {
      const renderer = create(
        <DecoratedText
          text={"Alpha\n\n\n\nBeta\n\n\n\n\n\n\nGamma"}
          metadataByChar={HashMap.empty()}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });

  describe("when receiving sophisticated text with different types of metadata and different blocks of newlines", () => {
    it("should render it correctly", () => {
      const renderer = create(
        <DecoratedText
          text={"Alpha1\nAlpha2\n\nBeta1\n我看书\nBeta3\nBeta4\n\n\n\nGamma"}
          metadataByChar={HashMap.of<Char, CharMetadata>(
            [
              "我",
              {
                annotation: "wǒ",
                className: Pinyin.fallingRising
              }
            ],
            [
              "看",
              {
                annotation: "kàn"
              }
            ],
            [
              "书",
              {
                className: Pinyin.flat
              }
            ]
          )}
        />
      );

      expect(renderer.toJSON()).toMatchSnapshot();
    });
  });
});
