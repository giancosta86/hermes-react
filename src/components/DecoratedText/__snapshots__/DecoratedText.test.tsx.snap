// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DecoratedText> when receiving sophisticated text with different types of metadata and different blocks of newlines should render it correctly 1`] = `
[
  <p>
    A
    l
    p
    h
    a
    1
    <br />
    A
    l
    p
    h
    a
    2
  </p>,
  <p>
    B
    e
    t
    a
    1
    <br />
    <ruby
      className="non-latin logogram chinese pinyin falling-rising"
    >
      我
      <rp>
        (
      </rp>
      <rt>
        wǒ
      </rt>
      <rp>
        )
      </rp>
    </ruby>
    <ruby>
      看
      <rp>
        (
      </rp>
      <rt>
        kàn
      </rt>
      <rp>
        )
      </rp>
    </ruby>
    <span
      className="non-latin logogram chinese pinyin flat"
    >
      书
    </span>
    <br />
    B
    e
    t
    a
    3
    <br />
    B
    e
    t
    a
    4
  </p>,
  <p>
    G
    a
    m
    m
    a
  </p>,
]
`;

exports[`<DecoratedText> when receiving text with leading and trailing space and no metadata should trim the text and render just plain characters in a single paragraph 1`] = `
<p>
  D
  o
  d
  o
</p>
`;

exports[`<DecoratedText> when receiving text with long sequences of intermediate newlines should ignore additional newlines 1`] = `
[
  <p>
    A
    l
    p
    h
    a
  </p>,
  <p>
    B
    e
    t
    a
  </p>,
  <p>
    G
    a
    m
    m
    a
  </p>,
]
`;

exports[`<DecoratedText> when receiving text with multiple intermediate carriageReturn + lineFeed should ignore carriage returns 1`] = `
<p>
  A
  l
  p
  h
  a
  <br />
  B
  e
  t
  a
  <br />
  G
  a
  m
  m
  a
</p>
`;

exports[`<DecoratedText> when receiving text with no newlines should just render it in a single paragraph 1`] = `
<p>
  我
  看
  <ruby
    className="non-latin logogram chinese pinyin flat"
  >
    书
    <rp>
      (
    </rp>
    <rt>
      shū
    </rt>
    <rp>
      )
    </rp>
  </ruby>
</p>
`;
